apps/emqx_retainer/src/emqx_retainer_mnesia.erl

```diff
@@ -281,18 +299,20 @@ do_store_retained_message(Msg, TopicTokens, ExpiryTime) ->
 
 do_store_retained_indices(TopicTokens, ExpiryTime) ->
     Indices = dirty_indices(write),
-    ok = emqx_retainer_index:foreach_index_key(
-        fun(Key) -> do_store_retained_index(Key, ExpiryTime) end,
-        Indices,
-        TopicTokens
-    ).
+    ok = mria:async_dirty(?RETAINER_SHARD, fun() ->
+        emqx_retainer_index:foreach_index_key(
```

Comment:
This will indeed will be more future-proof and concise, but I have to mention that this is unfair comparison. 🫠

Size of the first term is much larger than needed because the anonymous function is parsed by `erl_eval` (because we're in the shell) and not compiled into the module. And size of the second term is missing the external size of the argument list.

Reply:
One more thought. I'm honestly not sure if we want to make `emqx_retainer_index:foreach_index_key/3` essentially a part of application's RPC contract. This might turn out less future-proof actually.

Reply:
You are right, real difference would be much lower, sorry for confusion. :see_no_evil: